ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -52,4 +110,80 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
     }
     return details;
   }
+
+  /**
+   * A {@link Runnable} that cycles through the stores and executes compaction if required.
+   */
+  private class CompactionExecutor implements Runnable {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
+    private final ReentrantLock lock = new ReentrantLock();
+    private final Condition waitCondition = lock.newCondition();
+    private final Set<BlobStore> storesToSkip = new HashSet<>();
+    private final long waitTimeMs =
+        storeConfig.storeCompactionCheckFrequencyInHours * Time.MinsPerHour * Time.SecsPerMin * Time.MsPerSec;
+
+    /**
+     * Starts by resuming any compactions that were left halfway. In steady state, it cycles through the stores at a
+     * configurable frequency and runs compaction as required.
+     */
+    @Override
+    public void run() {
+      // complete any compactions in progress
+      for (BlobStore store : stores) {
+        if (store.isStarted()) {
+          try {
+            store.maybeResumeCompaction();
+          } catch (Exception e) {
+            logger.error("Compaction of store {} failed on resume. Continuing with the next store", store, e);
```

Comment:
Please note that the correctness and health of the store itself is not compromised here. The store is healthy and can work as before (it might have some duplicates because it was mid cycle and hard delete might be disabled). It is not in an unstable state to serve actual traffic i.e. there is no reason to shut the store down.